File: 剑指offer/剑指 Offer 13. 机器人的运动范围.cpp
//回溯法

class Solution {
public:
    int movingCount(int m, int n, int k) {
        if(k==0) return 1;
        vector<vector<bool>> valid(m,vector<bool>(n,true));
        return dfs(valid,m,n,0,0,k);
    }

    int dfs(vector<vector<bool>>& valid,int m,int n, int row , int col,int k){
        int sum=getSum(row)+getSum(col);
        if(row>=m||col>=n||sum>k||!valid[row][col]) return 0;
        valid[row][col]=false;
        return 1+dfs(valid,m,n,row+1,col,k)+dfs(valid,m,n,row,col+1,k);
    }

    int getSum(int x){
        int sum=0;
        for(;x>0;x/=10){
            sum+=x%10;
        }
        return sum;
    }
    
};

File: 剑指offer_java/剑指 Offer 13. 机器人的运动范围.java
class Solution {
    boolean [][]memory;
    int res=0;
    public int movingCount(int m, int n, int k) {
        if(k==0|| m==1&& n==1) return 1;
        memory=new boolean[m][n];
        dfs(0,0,k,m,n);
        return res;
    }
    void dfs(int x,int y,int k,int m,int n){
        if(x==m||y==n||memory[x][y]) return;
        if(gudge(x)+gudge(y)>k) return;
        memory[x][y]=true;
        res++;
        dfs(x+1,y,k,m,n);
        dfs(x,y+1,k,m,n);
    }
    int gudge(int x){
        if(x==100) return 1;
        return x/10+x%10;
    }
}
